File: orderService/controllers/order_controller.py
from flask import Blueprint, jsonify, request
from repositories.order_repository import *

order_blueprint = Blueprint('order_blueprint', __name__)
@order_blueprint.route('/create-order', methods=['POST'])
def create_order_controller():
    try:
        request_data = request.get_json()

        userID = request_data.get("userID")
        productIDs = request_data.get("productIds")
        quantities = request_data.get("quantities")
        totalPrice = request_data.get("total")
        addressID = request_data.get("shippingAddressId")
        paymentMethodID = request_data.get("paymentMethodId")

        result = add_order_to_db(userID, productIDs, quantities, totalPrice, addressID, paymentMethodID)
      
        if result:
            return jsonify({'result': 'Order added successfully'}), 201
        else:
            return jsonify({'result': 'Failed to add order'}), 500

    except Exception as e:
        return jsonify({'error': str(e)}), 500

@order_blueprint.route('/get-user-orders/<string:userID>', methods=['GET'])
def get_user_orders_controller(userID):
    try:
        order = get_user_order_by_id(userID)
        if order:
            return jsonify(order), 200
        else:
            return jsonify({'message': 'Order not found'}), 404

    except Exception as e:
        return jsonify({'error': str(e)}), 500
    
@order_blueprint.route('/update-order-status/<string:orderID>', methods=['PUT'])
def update_order_status_controller(orderID):
    try:
        request_data = request.get_json()
        status = request_data.get("status")

        result = update_order_status(orderID, status)
        return jsonify({'result': result}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500
    
@order_blueprint.route('/delete-order/<string:orderID>', methods=['DELETE'])
def delete_order_controller(orderID):
    try:
        result = delete_order_by_id(orderID)
        return jsonify({'result': result}), 200

    except Exception as e:
        return jsonify({'error': str(e)}), 500
    
@order_blueprint.route('/get-product-orders/<int:productID>', methods=['GET'])
def get_product_orders_controller(productID):
    try:
        order = get_product_orders(productID)
        if order:
            return jsonify(order), 200
        else:
            return jsonify({'message': 'Order not found'}), 404

    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: apiGateway/controllers/orderControllers/orderController.js
const express = require("express");
const router = express.Router();
const axios = require("axios");

async function createOrder(req, res) {
  try {
    const result = await axios.post("http://127.0.0.1:5001/create-order", req.body);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getOrder(req, res) {
  try {
    const orderID = req.params.orderID;
    const order = await axios.get(`http://127.0.0.1:5001/get-order/${orderID}`);
    res.json(order.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getAllOrders(req, res) {
  try {
    const orders = await axios.get("http://127.0.0.1:5001/get-orders");
    res.json(orders.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getUserOrders(req, res) {
  try {
    const userID = req.params.userID;
    const orders = await axios.get(`http://127.0.0.1:5001/get-user-orders/${userID}`);
    res.json(orders.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function updateOrderStatus(req, res) {
  try {
    const orderID = req.params.orderID;
    const status = req.body.status;
    const result = await axios.put(`http://127.0.0.1:5001/update-order-status/${orderID}`, req.body);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function deleteOrder(req, res) {
  try {
    const orderID = req.params.orderID;
    const result = await axios.delete(`http://127.0.0.1:5001/delete-order/${orderID}`);
    res.json(result.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

async function getProductOrders(req, res) {
  try {
    const productID = req.params.productID;
    const orders = await axios.get(`http://127.0.0.1:5001/get-product-orders/${productID}`);
    res.json(orders.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Internal Server Error" });
  }
}

module.exports = {
  createOrder,
  getOrder,
  getAllOrders,
  getUserOrders,
  updateOrderStatus,
  deleteOrder,
  getProductOrders,
};
